e2e-testing.md

```diff
@@ -0,0 +1,321 @@
+  - [4.1 One assertion per one test step method](#41-one-assertion-per-one-test-step-method)
+  - [4.2 Use expect for assertions](#42-use-expect-for-assertions)
+
+
+1\. Factory classes
+--------------
+
+#### 1.1. Use Factory classes only for Entities
+
+The ```FactoryItem``` class should be extended for each new Entity like User, Course, Event, StudentGroup class.
+For the tables which contains additional information for the Item it's recommended to create a ```link``` method.
+
+
+```typescript
+// ✅ good
+export class User extends FactoryItem {
+
+  public constructor(
+    apiClient: SeedAPI,
+    options: Options,
+  ) {
+    super(apiClient);
+    ...
+  }
+
+  async linkToCourse(
+    courseId: number,
+    status: CourseUserStatus,
+  ): Promise<void> {
+    await this.api.seedItemIfNotExist(
+      FactoryItemType.CourseUsers,
+      {
+        courseId,
+        userId: this.id,
+        status,
+      },
+    );
+  }
+}
+```
+
+2\. Page objects
+--------------
+
+#### 2.1. Define page locators outside the constructor
+
+Please define locators in the page object classes and components using the below example.
+There could be different approaches, but we agreed to follow a single style within the organization.
+
+```typescript
+// ❌ bad
+export class SignInPage extends BasePage{
+  private readonly emailField: Locator;
+
+  constructor(page: Page) {
+      super(page);
+      this.emailField = page.getByTestId('sign-in-user-email');
+  }
+}
+
+// ✅ good
+export class SignInPage extends BasePage{
+  private readonly emailField = this.page.getByTestId('sign-in-user-email');
+
+  constructor(page: Page) {
+    super(page);
+  }
+}
+```
+
+
+
+#### 2.2. Parametrized page URLs defining
+
+Use the below approach to define the parametrized page object URLs.
+
+```typescript
+// ✅ good
+interface Options {
+  chatId: number;
+}
+
+export class ChatPage extends LoggedInBasePage {
+
+  constructor(
+    page: Page,
+    options: Options,
+  ) {
+    super(page);
+
+    this.url = ROUTES.chat(options.chatId).index;
+  }
+}
+```
+
+
+#### 2.3. Use Components to re-use common page element groups
+
+
+If you defined some common elements from the header, footer, sideBar, or some popups that appear on several pages - make sure to define them in the corresponding component class.
+
+
+
+#### 2.4. Do not use dynamic values in allure step text
+
+
+Do not use dynamic values in allure step text becuase that way new test case will be added to the Allure TestOps on each test re-run.
+
+```typescript
+// ❌ bad
+async typeCourseName(name: string): Promise<void> {
+  await test.step(`Type course name ${name}`, async () => {
+    await this.courseDropwDown.type(name);
+  });
+}
+
+
+// ✅ good
+async typeCourseName(name: string): Promise<void> {
+  await test.step('Type course name', async () => {
+    await this.courseDropwDown.type(name);
+  });
+}
+
+```
+
+
+3\. Test scenarios
+--------------
+
+#### 3.1. Use snake_case for test folders naming
```

Comment:
why snake_case and not camelCase?
just asking, no need to change it

and ideally we setup linter rules

Reply:
Because we use _ to parse the Component & Feature names for Allure reports.